### ISSUE TITLE
Question about ES metric

### ISSUE THREAD

**Issue by sun1187** [locuslab/open-unlearning#106]
Hi, thanks again for your great work.

I have a question regarding the *extraction strength* metric.

In the original paper, both the equation and the code **regenerate the model's output while gradually increasing the prefix.**
![Image](https://github.com/user-attachments/assets/ac48aea0-0614-48e5-b115-6c8cccc83dae)
```
for _num_attempt_ in range(max([b - a for a, b in zip(left_bar, right_bar)])):
                    mid_bar = [(a + b) // 2 for a, b in zip(left_bar, right_bar)]
                    if _num_attempt_ != 0: 
                        if sum([int(l==r) for l, r in zip(mid_bar, old_mid_bar)]) == len(old_mid_bar): break
                    can_strings = token2string([tk[:cur] for cur, tk in zip(mid_bar, fuls_tks)])
                    inputs = tokenizer.batch_encode_plus(can_strings, add_special_tokens=True, return_tensors='pt', padding=True).to(model.device)
                    preds_tks = model.generate(inputs.input_ids, attention_mask=inputs.attention_mask, max_length=200, do_sample=False, use_cache=True, pad_token_id=tokenizer.eos_token_id)
                    _preds_tks_and_lens = string2token(tokenizer.batch_decode(preds_tks, skip_special_tokens=True))
                    pred_tks, pred_tks_lens = _preds_tks_and_lens['token'], _preds_tks_and_lens['length']
                    
                    pred_tks_ = [pred_tks[idx][mid_bar[idx]:len(fuls_tks[idx])] for idx in range(len(fuls_tks))]
                    fuls_tks_ = [fuls_tks[idx][mid_bar[idx]:] for idx in range(len(fuls_tks))]
                    if cfg.ps_type == 'similar':
                        match = [lcs(p,f) >= 0.5 * len(p) for p, f in zip(pred_tks_, fuls_tks_)]
                    else:
                        match = [sum([int(a == b) for a, b in zip(p, f)]) == len(p) for p, f in zip(pred_tks_, fuls_tks_)]
                    left_bar  = [left if match else mid  for match,  left, mid in zip(match,  left_bar, mid_bar)]
                    right_bar = [mid if match else right for match, right, mid in zip(match, right_bar, mid_bar)]
                    old_mid_bar = mid_bar
                ps_list += [1- (m-l)/(r-l) for l, m, r in zip(ques_tks_lens, right_bar, fuls_tks_lens)]
```

However, in your implementation, it seems that you **generate the output only once and then compare it with the reference answer to find the matching position.**
```
for log_probs, labels in zip(log_probs_batch, labels_batch):
            valid_len = len(labels)
            preds = torch.argmax(log_probs, dim=-1)
            for k in range(valid_len):
                suff_preds = preds[k:]
                suff_labels = labels[k:]
                if torch.equal(suff_preds, suff_labels):
                    break
            if valid_len == 0:
                # Rarely, tokenization can result in a mismatch with no valid target
                # tokens for loss computation (see preprocess_chat_instance() for
                # reference). Since this condition makes no sense in terms of
                # computing ES, we just choose to set ES=None
                logger.warning(
                    "ES score for an instance is marked None, due to "
                    "tokenization issues that resulted in no valid target tokens."
                )
                es_batch.append({"score": 0})
            else:
                es_score = 1 - (k / valid_len)
                es_batch.append({"score": es_score})
```
While both approaches ultimately identify the point of divergence between the answer and the reference, the implementation differs slightly.

**1. Was this setup chosen with a particular intention, such as improving efficiency or addressing computational constraints?
2. I'm also curious—does this change significantly affect the metric's performance or values compared to the original approach?**

Thank you for your help in advance!

**Comment by Dornavineeth** [comment]
Yes. Chosen for better efficiency. Ideally both the approaches should give exact same results because theoretically they are the same

**Comment by sun1187** [comment]
Thanks again for your quick response. @Dornavineeth 

I was wondering—since increasing the prefix length effectively changes the input to the model, I would expect the model’s response to change accordingly.

Could you please elaborate more on why the outputs are considered equivalent from a theoretical perspective?
I’d appreciate a bit more explanation on this point.

**Comment by molereddy** [comment]
We don’t generate token-by-token. Instead, we run one forward pass over the full input + target sequence and obtain next-token log-probs for every position.
Extraction strength is the smallest n such that, after feeding the first n target tokens, the arg-max prediction (next token generated in decoding) for every subsequent position equals the gold token. 
In practice, we find the last position where this arg-max alignment begins and use that as n.
This gives the identical result to the original input-decoding based method with much less compute.

All this assumes the standard greedy decoding
